server/diagram-api/src/main/java/io/typefox/sprotty/api/DefaultDiagramServer.java

```diff
@@ -156,27 +156,28 @@ public SModelRoot getModel() {
 	
 	@Override
 	public void setModel(SModelRoot newRoot) {
+		if (newRoot == null)
+			throw new NullPointerException();
 		synchronized(modelLock) {
-			if (newRoot == null)
-				throw new NullPointerException();
-			newRoot.setRevision(revision++);
+			newRoot.setRevision(++revision);
 			currentRoot = newRoot;
 		}
 		submitModel(newRoot, false);
 	}
 	
 	@Override
 	public void updateModel(SModelRoot newRoot) {
-		if (newRoot == null) {
-			submitModel(currentRoot, true);
-		} else {
-			synchronized(modelLock) {
+		synchronized(modelLock) {
```

Comment:
By moving synchronize up again, you have the lock during layout as well (which happens in submitModel)

Reply:
Sorry, I was wrong.